### DISCUSSION TITLE
Does not work cluster mode if I use single ip address of k8s?

### DISCUSSION THREAD

**Discussion by Park-Kwonsoo** [redis/rueidis#392] [Q&A] [answered]
Hello,

I am currently utilizing k8s to deploy and manage multiple Redis pods. In the current development environment, we have about three Redis pods running, all connected via a single endpoint. That is, within k8s, we have a host named 'redis'. When accessing this single endpoint, k8s handles the load balancing across the various Redis pods.

Therefore, when using rueidis, I've been specifying only the single endpoint in InitAddress. However, it seems there are occasional issues when invoking operations like HSet.

For debugging purposes, I utilized the Nodes method of rueidis's ClusterClient, and found that the number of Redis nodes being output is four instead of three. Among these, one appears to be registered with the endpoint address instead of the actual Redis IP.

To illustrate, if the actual IPs of the Redis pods are 1.2.3.4:6379, 1.2.3.5:6379, and 1.2.3.6:6379, and the endpoint set to access them is redis:6379, then when registering []host{"redis:6379"} in InitAddress, the keys in the map output from the client's Nodes are 1.2.3.4:6379, 1.2.3.5:6379, 1.2.3.6:6379, and redis:6379.

Could this be the reason certain operations like HSet are not functioning properly? I would greatly appreciate it if you could check this for me.

Thank you!

**Comment by rueian** [comment]
Hi @simonprickett,

The address appeared in the .Nodes() output is excepted and it is only used for refreshing shard information. So that can’t be the reason for certain operations that are not working properly.

What kind of errors you saw when doing HSet?

**Reply by Park-Kwonsoo** [reply]
Hello,

Firstly, I'd like to express my gratitude for your prompt feedback.

It seems that we occasionally encounter issues with HSet and HGet not setting or retrieving values properly. For instance, when we can't fetch a particular key (keyA) and field (fieldA) with HGet, we have logic to re-run HSet. This logic is executed multiple times in succession, leading to a situation where all related data gets lost in the process.

Initially, I thought this might be a synchronization issue arising from the client calling the logic too frequently. However, considering that this problem doesn't occur when using go-redis or only the single client of rueidis without any client changes, I began to suspect that there might be an issue with the driver itself.

Have there been any reports or discussions on similar topics or issues?

Thank you.

**Reply by rueian** [reply]
No. You are the first person reporting this.

Could you share the output of CLUSTER SHARDS command from all your redis nodes?

**Reply by Park-Kwonsoo** [reply]
Of course. Did you perhaps mean Cluster Nodes instead of Cluster Shards?

**Reply by Park-Kwonsoo** [reply]
Actually, we are still using Redis 6, not Redis 7, so that command is not supported.

**Reply by rueian** [reply]
Then please use CLUSTER SLOTS instead. It would be also helpful if you could provide code showing that how do you invoke HSet and HGet.

**Reply by Park-Kwonsoo** [reply]
Sure.

The result of ```CLUSTER SLOTS``` is
```
127.0.0.1:6379> cluster slots
1) 1) (integer) 0
   2) (integer) 5460
   3) 1) "10.202.64.136"
      2) (integer) 6379
      3) "8f93f8961e586b64db9469627a608736099bde14"
2) 1) (integer) 5461
   2) (integer) 10922
   3) 1) "10.202.102.152"
      2) (integer) 6379
      3) "69a7fb7c57666b5ced6fe99cdca413bd83632d26"
3) 1) (integer) 10923
   2) (integer) 16383
   3) 1) "10.202.138.243"
      2) (integer) 6379
      3) "5d252d4458a7ae08a143797ae9d841c5e67b91de"
```

and, I use `HSet` and `HGet` like this

```
func (r *Redis) HSet(ctx context.Context, key string, value any) (int64, error) {
	if r.delayedExecutionTimeLogging {
		checkTime := redisExecutionTimer("HSet", key)
		defer checkTime()
	}

	adapter := rueidiscompat.NewAdapter(r.client)
	cmd := adapter.HSet(ctx, key, value)
	return cmd.Val(), RedisErrorWrapf(cmd.Err(), "cache Key:%s", key)
}


func (r *Redis) HGet(ctx context.Context, key, field string) (string, error) {
	if r.delayedExecutionTimeLogging {
		checkTime := redisExecutionTimer("HGet", key)
		defer checkTime()
	}

	adapter := rueidiscompat.NewAdapter(r.client)
	cmd := adapter.HGet(ctx, key, field)
	return cmd.Val(), RedisErrorWrapf(cmd.Err(), "cache Key:%s", key)
}
```

**Reply by rueian** [reply]
Hi @Park-KwonSoo,

Thanks. That result of CLUSTER SLOTS looks normal to me.

Then what responses did you get from the `HSet` that were considered failures?
And what responses did you get from the `HGet` that were considered failures?

**Reply by Park-Kwonsoo** [reply]
Thank you for checking.
Firstly, there is no error message displayed when performing HSet. It seems to be working correctly.
The issue arises intermittently when performing HGet, resulting in a Redis.Nil error.
Upon directly connecting to Redis and using the redis-cli to query with HGet or Hgetall, I was able to confirm the presence of the respective key-field-value.
When I accessed each Redis pod and observed the incoming commands using the monitor command, there was no delete operation, like the del op, for that particular key.

**Reply by rueian** [reply]
Hi @Park-KwonSoo, could you observe the problematic HGET/HGETALL from the server side using the monitor command? I know this is almost impossible but for now, I don't have more information for further debugging.

By the way, do all your redis nodes have the same response of `cluster slots`?

**Reply by rueian** [reply, marked as answer]
Hi @Park-KwonSoo, is there any update?

**Reply by Park-Kwonsoo** [reply]
Oh sorry, I missed your response while I was looking into another issue. To get straight to the point, there were no changes made, but the issue isn't occurring anymore. My guess is that either the Redis we're using had a temporary glitch, or there was some network issue, or the client was making an incorrect call - one of those three, but I'm not certain. I tried various methods for debugging, but the issue stopped occurring before I could accurately pinpoint it, so I've stopped tracking it for now.
Thank you so so much for your help, and I really appreciate you providing the useful driver!
Have a great day :)
